ring/client/pool.go

```diff
@@ -173,24 +179,28 @@ func (p *Pool) removeStaleClients() {
 	}
 }
 
-// cleanUnhealthy loops through all servers and deletes any that fails a healthcheck.
+// cleanUnhealthy loops through all servers and deletes any that fail a healthcheck.
+// The health checks are executed concurrently with p.cfg.MaxConcurrentHealthChecks.
 func (p *Pool) cleanUnhealthy() {
-	for _, addr := range p.RegisteredAddresses() {
+	addresses := p.RegisteredAddresses()
+	_ = concurrency.ForEachJob(context.Background(), len(addresses), p.cfg.MaxConcurrentHealthChecks, func(ctx context.Context, idx int) error {
+		addr := addresses[idx]
 		client, ok := p.fromCache(addr)
 		// not ok means someone removed a client between the start of this loop and now
 		if ok {
-			err := healthCheck(client, p.cfg.HealthCheckTimeout)
+			err := healthCheck(ctx, client, p.cfg.HealthCheckTimeout)
 			if err != nil {
 				level.Warn(p.logger).Log("msg", fmt.Sprintf("removing %s failing healthcheck", p.clientName), "addr", addr, "reason", err)
 				p.RemoveClientFor(addr)
 			}
 		}
-	}
+		return nil
+	})
 }
 
 // healthCheck will check if the client is still healthy, returning an error if it is not
-func healthCheck(client PoolClient, timeout time.Duration) error {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+func healthCheck(ctx context.Context, client PoolClient, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
```

Comment:
The context could not be cancelled elsewhere. But when looking at the `concurrency.ForEachJob` function again, I realised, that it would stop after the first error. This would prevent it from executing all health checks, and that's definitely not what we want.

Reply:
@bboreham I implemented the concurrency directly in the `cleanUnhealthy()` function.

Reply:
> The context could not be cancelled elsewhere. But when looking at the concurrency.ForEachJob function again, I realised, that it would stop after the first error. This would prevent it from executing all health checks, and that's definitely not what we want.

Not sure reimplementing it from scratch (as done in the last commit) is a good approach. Why don't we keep using `concurrency.ForEachJob()` but we never return an error from the function? That's how we use it in Mimir when we don't want to stop on first error.

Reply:
Reverted the last commit.